### DISCUSSION TITLE
On/Off Peak are a 0-70 seconds after the switch

### DISCUSSION THREAD

**Discussion by thewishy** [BottlecapDave/HomeAssistant-OctopusEnergy#420] [General]
Hi, 

Wasn't really sure how best to raise this, maybe a feature request, maybe a bug, maybe I'm using it wrong. Currently using v8.2.0. I've got IO, and a home battery system (Givenergy) which I want to align with the off peak time so that 
1) I get to charge up during off peak periods
2) If there are out of normal window charging sessions, I don't dump the battery into them.

I've tied the On/Off peak sensor to an automation to ask the GivEnergy system to turn on/off consumption balancing. It's been working reasonably well. It's just rough edges which are costing me maybe a 0.05 kWh each day. 

Crux of it is that the sensor doesn't switch off until anything up to a minute or so after the window has ended. My system isn't generally CPU constrained, I doubt it's waiting to execute for an extended period of time. More likely HA evaluates vs system clock every minute. 

24 September 2023
Octopus Electric Off Peak turned off
07:31:04 - 8 hours ago (Extended window sue to charging)
23 September 2023
Octopus Electric Off Peak turned on
23:30:20 - 16 hours ago
Octopus Electric Off Peak turned off
05:30:49 - Yesterday (Typical day)

Looking at GivTCP, the changes to battery settings seem to line up, so it doesn't seem to be a delay with the recorder
2023-09-24 07:31:04,952 - Inv1 - write       -  [INFO    ] - Setting dayRate via external trigger
2023-09-24 07:31:04,994 - Inv1 - mqtt_client -  [ERROR   ] - Force Charge is not currently running
2023-09-24 07:31:05,305 - Inv1 - write       -  [INFO    ] - Setting demand mode was a success

Possible fixes
1) Expose or hard set offset, so that I can ask on-peak events to trigger a minute early, which means we're always going off grid a smidge before the smart-meter clock ticks over
2) Expose evaluation function, which I can automate to happen exactly on the half hour

**Comment by BottlecapDave** [comment]
Hi there. Yes unfortunately this is tied to HA and its update schedule. The entities are currently set to evaluate every minute as I didn't want my integration to be the cause of peoples HA instances grinding to a hault where 40+ sensors all evaluated frequently. This was also driven when a lot of the data was fed from the sensors themselves instead of a central update system. 

I'll see if I can reduce this evaluation period down for the future, but I guess there's still a chance it could go over by tens of seconds. However this scenario is possible even without HA with inverter clocks and smart meters clocks being out slightly. There is also no promises this would even be possible.

For now, assuming you're usng continuous target rate sensor, you should be able to create a template sensor which exposes the proposed endtime minus any off set

```
{{ state_attr('binary_sensor.octopus_energy_target_test', 'target_times')[:-1][0]['valid_to'] - timedelta(minutes=1) }}
```

You could then use this sensor as an additional trigger when it goes into the past.

You _might_ also be able to use the [update_config](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_target_rate.md#service-octopus_energyupdate_target_config) service with an identical setting (e.g. target hours), which should trigger a re-evaluation if I understand the HA sensor lifecycle correctly.

**Reply by thewishy** [reply]
Thanks, that's helpful. 

So the home assistant box is synced with NTP, and we've also got a GPS stratum 1 timesource running off a pi in the loft. And the smartmeter seems to have a reliable timesource, as they're within a second of each other. 
The signalling to the inverter is "active" from home assistant via GivTCP (I basically have a long running automation which extends the window periodically, and then when we switch to peak it sends a cancel forced charge message) - seems to take about half a second to action the configuration change. So that aspect is addressed at least.

I think, in general, most people would prefer that the off peak window finish up to a minute early rather than a minute late, but depends on peoples automations

I'm just using the Off Peak sensor at the moment, but I should be able to puzzle my way through a template sensor

**Reply by BottlecapDave** [reply]
Sorry, I misread the sensor you were using - the template sensor isn't going to work with the `is_off_peak` sensor :( 

The template sensor I mentioned was for a [target rate](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/setup_target_rate.md) sensor, which you could create and configure to match the off peak timeframe for your tariff. This would then expose the required data for the template sensor.

> I think, in general, most people would prefer that the off peak window finish up to a minute early rather than a minute late, but depends on peoples automations

If I manage to get the sensor to evalate more frequently, I'll take this into consideration. I'm trying not to have too many configurations as it complicates the integration and makes things very messy. Another option might be to expose more information via attributes so if it doesn't meet people's needs then they can build off the data at the very least.

**Comment by thewishy** [comment]
Some progress with tinkering. The following automation allows for a forced update at the correct times. The mode switch seems to happen between 0 and 10 seconds later, which is better than 0-70 seconds without. 
```
alias: Octopus - update off peak
description: ""
trigger:
  - platform: time_pattern
    minutes: "30"
  - platform: time_pattern
    minutes: "0"
condition: []
action:
  - service: homeassistant.update_entity
    data: {}
    target:
      entity_id: >-
        binary_sensor.octopus_energy_electricity_***_***_off_peak
mode: single
```

**Comment by thewishy** [comment]
Spending a little longer looking into this. 
Code to change state a little before the change
```
def is_on(self):
    """Determine if current rate is off peak."""
    if (self._state):
      # Off Peak, check if we'll be back to on-peak in the next minute
      current = now()+timedelta(minutes=1)
    else:
     # On peak, check if we're now off-peak
     current = now()
```

(Indentation is almost certainly screwed up there...)

A few questions
- Does a default of ending the off peak a smidge early rather than a smidge late make sense to you?
- Does this seem sane as an approach?
- Would you like me to have a go at a pull request, or is the code change trivial enough?

**Reply by BottlecapDave** [reply]
The code change is trivial enough, I just fear I'll get people raising bug reports stating it ends too early instead of too late (you can rarely win in software development ;)) But I take your point that a bit too early is better than a bit too late.

I'll need to think about it...

**Reply by thewishy** [reply]
I understand.

How about an alternative - have a next session off peak sensor, and then at :00 and :30 it's just a simple "if current slot is offpeak, and next slot is on peak then go go gadget batteries" And you could chose to do that 29:30 or whatever, it's users choice from there...
